### PR TITLE
8336097: UserAgent Styles using lookups are promoted to Author level if look-up is defined in Author stylesheet

### PR DESCRIPTION
**Pull request by hjohn**
This change removes the origin determination from `resolveLookups`.  Instead, the origin from the style is used.

Although a comment in the code alluded that this may cause problem with `INLINE` styles, this is not the case. Whenever a `Node` is associated with a `CssStyleHelper`, a suitable shared cache is determined for its use.  This already takes into account the presence of an inline style, and only nodes with the same inline style can share such a cache.  See `Cache#getStyleMap` and specifically this fragment where an additional selector is added for the inline style:

            if (hasInlineStyle) {
                Selector selector = cacheContainer.getInlineStyleSelector(inlineStyle);
                if (selector != null) selectors.add(selector);
            }

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8336097](https://bugs.openjdk.org/browse/JDK-8336097): UserAgent Styles using lookups are promoted to Author level if look-up is defined in Author stylesheet (**Bug** - P3)

### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Michael Strauß](https://openjdk.org/census#mstrauss) (@mstr2 - Committer)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * @Jahin9999 (no known openjdk.org user name / role)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1503/head:pull/1503` \
`$ git checkout pull/1503`

Update a local copy of the PR: \
`$ git checkout pull/1503` \
`$ git pull https://git.openjdk.org/jfx.git pull/1503/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1503`

View PR using the GUI difftool: \
`$ git pr show -t 1503`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1503.diff">https://git.openjdk.org/jfx/pull/1503.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1503#issuecomment-2225298405)

### CODE REVIEW

**Comment by andy-goryachev-oracle** [comment]
Would it be possible to create a set of tests that verify _all_ the possible combinations of priorities and stylesheets (as well as Application/Scene/Parent/inline/code)?

Also, do you think cssref.html should be updated to clarify the behavior in "Scene, Parent and SubScene Stylesheets" section?

**Comment by hjohn** [comment]
> Would it be possible to create a set of tests that verify _all_ the possible combinations of priorities and stylesheets (as well as Application/Scene/Parent/inline/code)?
> 
> Also, do you think cssref.html should be updated to clarify the behavior in "Scene, Parent and SubScene Stylesheets" section?

I've added a test for the specific fix I did.  Without my fix, cases 5 and 6 (`PROPERTY_OVERRIDES_USER_AGENT_VARIABLE_SET_IN_AUTHOR` and `PROPERTY_OVERRIDES_USER_AGENT_VARIABLE_SET_INLINE`) fail. With my fix, they pass.

**Comment by hjohn** [comment]
> Also, do you think cssref.html should be updated to clarify the behavior in "Scene, Parent and SubScene Stylesheets" section?

@andy-goryachev-oracle 
That section is about cascading rules, and doesn't even mention lookups.  In fact, the whole CSS reference does not mention lookups (aside from mentioning that JavaFX has a "property lookup" feature, and an indirect mention in the form of "looked-up-color").

There is no mention how property look ups work (and what types it supports), how they should be named (any name is allowed as long as it is not a supported color name), or how they cascade (before this fix, they gave styles a new priority, after this fix, the priority of the style is unchanged).

As this is quite some work, I documenting this feature should be a separate task.

**Comment by andy-goryachev-oracle** [comment]
> As this is quite some work, I think documenting this feature calls for a separate task to avoid delaying this fix.

Yes, it's just a question - I don't want to delay this PR.  I asked because you do write good docs.

**Comment by kevinrushforth** [comment]
The fix looks good, presuming that you are correct, and that the comment about needing to set the origin of the  parsed value to the greatest of any resolved reference in order to avoid storing an inline style in the shared cache is in fact wrong. Testing should help catch any potential problems.

The new unit test looks like it covers all the cases. As with PR #1505, this PR mixes JUnit 4 and JUnit 5, which we really don't want to do in the same test class. Consider converting the entire class to JUnit 5, perhaps in #1505.

**Comment by kevinrushforth** [comment]
> The new unit test looks like it covers all the cases. As with PR #1505, this PR mixes JUnit 4 and JUnit 5, which we really don't want to do in the same test class. Consider converting the entire class to JUnit 5, perhaps in #1505.

On closer look, it's probably best to not do that as part of #1505 since that PR only adds a couple test methods to an existing class with many preexisting tests. Similarly, this PR only adds one new test method.

So I think the best options are:

1. File a new test bug and PR that converts this class to JUnit 5, that new bug would block both of the other bugs
2. Stick with JUnit 4 for both of the existing PRs

Comments?

**Comment by hjohn** [comment]
> > The new unit test looks like it covers all the cases. As with PR #1505, this PR mixes JUnit 4 and JUnit 5, which we really don't want to do in the same test class. Consider converting the entire class to JUnit 5, perhaps in #1505.
> 
> On closer look, it's probably best to not do that as part of #1505 since that PR only adds a couple test methods to an existing class with many preexisting tests. Similarly, this PR only adds one new test method.
> 
> So I think the best options are:
> 
> 1. File a new test bug and PR that converts this class to JUnit 5, that new bug would block both of the other bugs
> 2. Stick with JUnit 4 for both of the existing PRs
> 
> Comments?

Perhaps:

3. Put the JUnit 5 code in a separate test?

About mixing asserts; I meant to upgrade all of them in this PR, but I missed a few.  Mixing asserts however is technically not an issue as both frameworks will work with each others asserts (ie. JUnit 4 can use JUnit 5 asserts, or even asserts from a 3rd party library like AssertJ).  What you can't mix is the annotations.  You must use all JUnit 4 or JUnit 5 annotations (and this I did do correctly).

I would personally prefer not to write more "old" JUnit 4 code than necessary, especially for parameterized cases or testing exceptions (JUnit 4 just is showing its age there), so if full conversion to JUnit 5 is not an option, I'll file a ticket to first convert it.

**Comment by hjohn** [comment]
> The fix looks good, presuming that you are correct, and that the comment about needing to set the origin of the parsed value to the greatest of any resolved reference in order to avoid storing an inline style in the shared cache is in fact wrong. Testing should help catch any potential problems.

I did some local testing as well regarding this (before I wrote the test cases) to first see if this fix would not break anything.  I couldn't get inline styles that mess with `-fx-base` to "poison" each other.  I also went through the caching logic, and found that any difference in inline styles results in a separate cache space -- this makes sense, because if inline styles were not taken into account when selecting the cache, then any inline style for two otherwise equal nodes could poison the other -- in other words, inline styles would not work correctly when cached if these were not taken into account, lookups or not.

**Comment by hjohn** [comment]
> I know it might be quite a bit of work, but I think it might be worth going through an exercise to enumerate all possible scenarios in a test.
> 
> As a reviewer, I cannot begin to figure out the possible consequences of the change in this PR without documenting the rules first.

The CSS rules, or the rules surrounding lookups?

For lookups:

**Before**

If a style used a variable, and that variable was overridden in a stylesheet with a higher priority `StyleOrigin`, then the original style would be modified to have this higher priority as well.  If there was a chain of references (ie. `-fx-fill-color` refers to `-fx-base`) then the highest priority encountered while resolving the variable would be used.  It is possible that an inline style overrides a variable like `-fx-base`, which would be the highest `StyleOrigin` priority, and so the original style would get `INLINE` priority.  There was a worry that this would badly interact with caching (as inline styles are supposed to be local to the `Node` and so by definition, can't be shared with any other nodes), however, such a separation in caches was already present (out of necessity) was this would apply to any inline style (like `-fx-text-fill: red`), whether or not it used a lookup.  The separation is therefore always present when an inline style is present, irrespective of the entire variable lookup system.

**After**

A lookup variable does not alter the priority of the style referring the variable in any way.  This both simplifies the code, as well as the reasoning about your style priorities.

**Potential Consequences**

First realize that this is limited in scope to that variable lookups in JavaFX allow.  They only allow color replacements.  Also realize that the combination of setting some colors with a setter, while adjusting others by changing style sheet variables like `-fx-base` is a rare one.  Most often projects will mandate it is one or the other (ie. specify all colors in stylesheets, or don't use stylesheets at all).

1. Leaving it as-is: versions of JavaFX prior to 21 behaved differently in this matter (due to a bug), and people have noticed that in 21+ FX is now overriding color setters as if some styles are originating from higher priority stylesheets.  This is limited in scope to colors, but could get worse if the system is extending to allow for other uses of variables.

2. Modifying the (undocumented) behavior to be more in line with what one would expect: due to a seemingly unrelated bug, versions before JavaFX 21 often didn't override setter values (all of them, in specific situations, not just colors).  This was mostly expected as both the call to the setter and the (higher priority) author stylesheets are under control of the developer, and so if there was any conflict, it could be solved by the developer.  In other words, developers could fix their own mistake, even though it may have been a bug on our side.

TLDR; modifying the behavior will make JavaFX operate more like it did prior to JavaFX 21.  None of them however will work exactly the same in all situations (ie. JavaFX 20 works differently from 21, and a future version where this PR merged will work differently again.  However, that future version will work closer to how JavaFX behaved before version 21.

Maybe a table will help...

First, the order form highest to lowest priority is: `INLINE` (using `Node::setStyle`), `AUTHOR` (from user set stylesheets), `USER` (for setters) and then `USER_AGENT` (a stylesheet like Modena set by FX itself).

| Action \ Version |<21|21+|With Fix|
|---|---|---|---|
|Setting a color|Not overridden by anything(*)|Overridden by AUTHOR and INLINE styles, and by any USER_AGENT styles if they use a variable and that variable was specified in an AUTHOR or INLINE style|Only overridden by AUTHOR or INLINE styles (as documented)|
|Setting a non-color|Not overridden by anything(*)|Only overriden by AUTHOR or INLINE styles (as documented)|Same as 21+|
|Using `-fx-base` in AUTHOR stylesheet, as well as some direct styles|Both USER_AGENT styles that use `-fx-base` and AUTHOR styles that apply things directly would be in competition who is the most specific|Same as before 21|USER_AGENT styles that use `-fx-base` are not in direct competition with directly set AUTHOR styles|

(*) most of the time, the situation is a bit complicated due to cache sharing

**Comment by andy-goryachev-oracle** [comment]
> First, the order from highest to lowest priority is: `INLINE` (using `Node::setStyle`), `AUTHOR` (from user set stylesheets), `USER` (for setters) and then `USER_AGENT` (a stylesheet like Modena set by FX itself).

What are the priorities between

- system stylesheet (modena.css) set via `StyleManager.getInstance().setUserAgentStylesheets()`
- user agent stylesheet added via `Application.setUserAgentStylesheet()`
- user agent stylesheet added via `Scene.setUserAgentStylesheet()`
- user agent stylesheet added via `SubScene.setUserAgentStylesheet()`

**Comment by hjohn** [comment]
> > First, the order from highest to lowest priority is: `INLINE` (using `Node::setStyle`), `AUTHOR` (from user set stylesheets), `USER` (for setters) and then `USER_AGENT` (a stylesheet like Modena set by FX itself).
> 
> What are the priorities between
> 
> * system stylesheet (modena.css) set via `StyleManager.getInstance().setUserAgentStylesheets()`
> * user agent stylesheet added via `Application.setUserAgentStylesheet()`
> * user agent stylesheet added via `Scene.setUserAgentStylesheet()`
> * user agent stylesheet added via `SubScene.setUserAgentStylesheet()`

Here is the order that stylesheets are added to a style map, from how `StyleManager::findMatchingStyles` works.  The order is from lowest to highest precedence:

## Primary user agent stylesheet:

One of these (only one, first one wins):

- `SubScene` user agent stylesheet
- `Scene` user agent stylesheet (**only** if there was no `SubScene` user agent stylesheet)
-  Application user agent stylesheet **only** if there was no `SubScene` or `Scene` user agent stylesheet (in other words, `Scene` and `SubScene` user agent stylesheets will completely replace modena)

## Additional `Region` user agent stylesheet

Regions can provide a user agent stylesheet by overriding `Region::getUserAgentStyleSheet`.  Only one is added, from the nearest parent, and it does **not** replace the primary user agent stylesheet.

## Author stylesheets, from lowest to highest precedence:

- From `Scene`
   - If a Scene has multiple stylesheets, the first/last stylesheet in the list has lowest/highest precedence respectively.
- Then `Parent` stylesheets
   - The least specific parent's stylesheets (closest to the root) have lowest precedence
   - If a Parent has multiple stylesheets, the first/last stylesheet in the list has lowest/highest precedence respectively.

# Style selection

Styles are grouped per style origin, from lowest to highest precedence:

- User Agent
- Author
- Inline

A style from within a higher precedence group will always be applied first.

Styles within the same style origin group will use specificity (how specific the style is) to determine which should be applied first, with the most specific style taking precedence.

If there are styles with the same specificity, then the style from the nearest `Node`, `Parent` or `Scene` takes precedence (this is determined by the order the stylesheets are managed by `StyleManager` explained in the first part).

If there is still a conflict (ie. two styles with the same precedence, both coming from the same `Node`) then the latest stylesheet added takes precedence.

**Comment by andy-goryachev-oracle** [comment]
Thank you for the detailed writeup!  I did not know about `Region.getUserAgentStylesheet()`, especially about its doc "Some JavaFX CSS implementations may choose to cache this response for an indefinite period of time, and therefore there should be no expectation around when this method is called."  How are we supposed to implement dynamically changing styles??

Do I understand you correctly that `Scene.setUserAgentStylesheet()` completely replaces modena.css for that Scene?  And then Scene.getStylesheets() will lay on top of that, having higher priority?

**Comment by hjohn** [comment]
> Thank you for the detailed writeup! I did not know about `Region.getUserAgentStylesheet()`, especially about its doc "Some JavaFX CSS implementations may choose to cache this response for an indefinite period of time, and therefore there should be no expectation around when this method is called." How are we supposed to implement dynamically changing styles??

The `Region` based system doesn't look well thought out, but I suppose you can force a reload of all styles by changing something at a higher level (like `Scene`).  I didn't know it existed either, and it seems to behave differently to other user agent stylesheets (with Application / Scene / SubScene, only one applies, while `Region` can complement it further...) -- I'm guessing this is intended for 3rd parties, so they can style their controls without those styles competing at the same level as author stylesheets (ie. they're sort of "extending" modena).

As for dynamically changing styles, it depends on who wants those.  If it is the control developer, they can ensure their stylesheet has multiple variants and use a top level selector for changing between one or the other (for example between light and dark).

If it is the user of the control, they can do so normally by just replacing stylesheets (or preferably, also using a selector, as that's much faster).

> Do I understand you correctly that `Scene.setUserAgentStylesheet()` completely replaces modena.css for that Scene? And then Scene.getStylesheets() will lay on top of that, having higher priority?

Yes, that's correct.  The code will not use the platform user agent stylesheet when there is either a scene or subscene user agent stylesheet available.  This seems to make it possible to say have a Scene in modena style, and another Scene in caspian style without the Application stylesheet potentially intefering.

And yes, Scene.getStylesheets layers on top of all that (but those are author stylesheets, so they overrule the user agent stylesheets anyway).

**Comment by andy-goryachev-oracle** [comment]
thanks for the writeup!  these are quite helpful.

since this PR changes the way CSS is resolved, I wonder if we need a CSR.

**Comment by kevinrushforth** [comment]
> since this PR changes the way CSS is resolved, I wonder if we need a CSR.

That's a good question. My take is that we don't, since this is a bug fix that restores the behavior to what it was before a previous bug fix exposed the bug.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java [hunk 1]
```diff
@@ -60,7 +70,7 @@ private static void resetStyleManager() {
         sm.hasDefaultUserAgentStylesheet = false;
     }
 
-    @Before
+    @BeforeEach
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java:73]: since we are now creating Stage every time, shouldn't we `hide()` it each time?

**Review by andy-goryachev-oracle — COMMENTED** [review]
Thank you for all the explanations!

I did a quick test with the Monkey Tester, setting various styles.  Haven't found any issues.

Left a few comments inline.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java (hunk 1 above)
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java:73]: This was already done each time, it has not changed, only the annotation.  The semantics of `@Before` in JUnit 4 and `@BeforeEach` in JUnit 5 are the same.  However, see #1515 as I'll remove these changes from this PR.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java [hunk 2]
```diff
@@ -658,4 +668,90 @@ public void initialNodeWithUserSetValueShouldNotResetValuesOnOtherNodesWithoutOv
...
+
+        // Direct inline styles always win:
+        INLINE_OVERRIDES_USER_AGENT(RED_STYLESHEET, null, null, "-fx-background-color: #808080"),
+        INLINE_OVERRIDES_PROPERTY(RED_STYLESHEET, Color.BLUE, null, "-fx-background-color: #808080"),
+        INLINE_OVERRIDES_AUTHOR(RED_STYLESHEET, null, RED_STYLESHEET, "-fx-background-color: #808080"),
+
+        // Indirect inline styles win when not directly overridden by author stylesheet or property:
+        INLINE_VARIABLE_OVERRIDES_USER_AGENT_VARIABLE(RED_INDIRECT_STYLESHEET, null, null, "-fx-base: #808080"),
+        INLINE_VARIABLE_OVERRIDES_AUTHOR_VARIABLE(RED_INDIRECT_STYLESHEET, null, FX_BASE_GREEN_STYLESHEET, "-fx-base: #808080");
+
```
[the remark this answers is not in the document]
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java:710]: > Another question: do we want to go through all the cases of Scene/SubScene/Parent/Region user agent style sheet, for completeness sake, or would that be a different test file?

In my opinion, this does seem a bit of out of scope for this change.  it would for example not add any new coverage to the code that was changed.  Also the origin of the user agent stylesheet is not a factor in the resolution of lookups.  The user agent stylesheets, their replacement (via Scene/Subscene) or addition (via Region) should (already?) be tested elsewhere.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java (hunk 1 above)
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java:73]: I'm not entirely sure on the semantics of `Stage` when it is run with the stub toolkit; I suspect they won't stick around when no longer referenced, or this test would have been an issue when it was created years ago already.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java [hunk 3]
```diff
@@ -658,4 +668,90 @@ public void initialNodeWithUserSetValueShouldNotResetValuesOnOtherNodesWithoutOv
...
+        AUTHOR_OVERRIDES_PROPERTY(RED_STYLESHEET, Color.BLUE, GRAY_STYLESHEET, null),
+        AUTHOR_OVERRIDES_USER_AGENT_VARIABLE_SET_INLINE(RED_INDIRECT_STYLESHEET, null, GRAY_STYLESHEET, "-fx-base: yellow"),
+
+        // Indirect author styles win when property is not set directly, and there is no direct or indirect override in an inline style:
+        AUTHOR_VARIABLE_OVERRIDES_USER_AGENT_VARIABLE(RED_INDIRECT_STYLESHEET, null, FX_BASE_GRAY_STYLESHEET, null),
+
+        // Direct inline styles always win:
+        INLINE_OVERRIDES_USER_AGENT(RED_STYLESHEET, null, null, "-fx-background-color: #808080"),
+        INLINE_OVERRIDES_PROPERTY(RED_STYLESHEET, Color.BLUE, null, "-fx-background-color: #808080"),
+        INLINE_OVERRIDES_AUTHOR(RED_STYLESHEET, null, RED_STYLESHEET, "-fx-background-color: #808080"),
```
[the remark this answers is not in the document]
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java]: I've added cases where user agent stylesheet is `null`, but I've not added all combinations -- I feel there is no point to check if an inline style will override all combinations of user agent, property and author styles when there are already 3 cases that check if the inline style will override user agent, property and author styles individually.  That said, I've added a few cases where ALL competing options are set (ie. for inline, a case where UA, property and author are set, and see if inline still wins).

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java [hunk 4]
```diff
@@ -658,4 +668,90 @@ public void initialNodeWithUserSetValueShouldNotResetValuesOnOtherNodesWithoutOv
...
+        // Property wins when not directly overridden by author or inline style:
+        PROPERTY_OVERRIDES_USER_AGENT(RED_STYLESHEET, Color.web("#808080"), null, null),
+        PROPERTY_OVERRIDES_INDIRECT_USER_AGENT(RED_INDIRECT_STYLESHEET, Color.web("#808080"), null, null),
+        PROPERTY_OVERRIDES_USER_AGENT_VARIABLE_SET_IN_AUTHOR(RED_INDIRECT_STYLESHEET, Color.web("#808080"), FX_BASE_GREEN_STYLESHEET, null),
+        PROPERTY_OVERRIDES_USER_AGENT_VARIABLE_SET_INLINE(RED_INDIRECT_STYLESHEET, Color.web("#808080"), null, "-fx-base: yellow"),
+
+        // Author style wins when not directly overridden by inline style:
+        AUTHOR_OVERRIDES_USER_AGENT(RED_STYLESHEET, null, GRAY_STYLESHEET, null),
+        AUTHOR_OVERRIDES_INDIRECT_USER_AGENT(RED_INDIRECT_STYLESHEET, null, GRAY_STYLESHEET, null),
+        AUTHOR_OVERRIDES_PROPERTY(RED_STYLESHEET, Color.BLUE, GRAY_STYLESHEET, null),
```
[the remark this answers is not in the document]
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java]: I've added the indirect cases in a separate blocks.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java:714]: I added a few cases where user agent is `null`.  I've not added all combinations as that would lead to 20-30 extra cases that add little of value.  If I already verified that INLINE will override the lower priority sources individually, then it's safe to say that all possible other combinations of those lower priority sources will also get overridden.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java (hunk 1 above)
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java]: The annotation changes have now been removed from this PR now that #1515 has been integrated.

**Review by mstr2 — APPROVED** [review]
I've tested a few cases, which work as expected. This patch looks good to go.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java:714]: I still think there is a value in providing exhaustive coverage, especially since the number of combination is relatively low.  The added value, for instance, is to guard against regressions introduced by other people.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java (hunk 1 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java]: This question may not relate to this PR as it does not change the existing logic.

I wonder if the lingering Stages affect the test(s) in some way.  Is FX subsystem getting shut down after each test class?  Does anyone know?

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java [hunk 5]
```diff
@@ -657,4 +663,99 @@ public void initialNodeWithUserSetValueShouldNotResetValuesOnOtherNodesWithoutOv
...
+        // User agent styles win when not overridden directly or indirectly:
+        UA(GRAY_STYLESHEET, null, null, null),
+        INDIRECT_UA(GRAY_INDIRECT_STYLESHEET, null, null, null),
+
+        // Property wins when not directly overridden by author or inline style:
+        PROPERTY(null, Color.web("#808080"), null, null),
+        PROPERTY_OVERRIDES_UA(RED_STYLESHEET, Color.web("#808080"), null, null),
+        PROPERTY_OVERRIDES_INDIRECT_UA(RED_INDIRECT_STYLESHEET, Color.web("#808080"), null, null),
+
+        // Property wins even if indirectly overridden by author or inline style (resolving of a lookup does not change priority of the user agent style)::
```
Comment (mstr2) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java:689]: You might want to remove the double colon at the end of the sentence.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java:714]: @andy-goryachev-oracle 

Could you be more specific which cases (related to this change and lookups) are currently not being adequately covered?
  
I'm of the opinion that the tests are more than sufficient for the changes involved, I don't think it is the responsibility of this PR to add coverage in areas unrelated to it.  The cases you are requesting IMHO check cascading rules, which should already be verified elsewhere (and if not, is certainly not something that should be part of this PR).

The tests I added include 19(!) new cases (versus the 13 that existed before), but don't add **any** new coverage (lookups were already covered by for example `removingThenAddingNodeToDifferentBranchGetsIneritableStyle`) -- they're only adding assertions specifically for this change.  One might even say I overdid it here, as half of the tests don't even deal with lookups (ie. all the cases where no "indirect" variant is involved). Making these test cases exhaustive for something that has 4 variables and 3-4 options each would result in 80+ cases.

So adding further tests does not seem to serve any purpose from where I stand; most of the requested additional cases would not even involve lookups, or would just confirm existing cascading rules (if an inline style overrides a property **or** author style, then it would also override property **and** author style). And none of them would extra coverage...

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java (hunk 5 above)
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java]: Fixed, thanks

**Review by mstr2 — APPROVED** [review]

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java:714]: fine

**Review by andy-goryachev-oracle — APPROVED** [review]

**Review by kevinrushforth — APPROVED** [review]
Looks good.
